lib/src/strings.dart

```diff
@@ -94,38 +95,100 @@ String _tryYamlEncodeSingleQuoted(String string) {
 /// It is important that we ensure that [string] is free of unprintable
 /// characters by calling [_hasUnprintableCharacters] before invoking this
 /// function.
-String _tryYamlEncodeFolded(String string, int indentation, String lineEnding) {
-  String result;
-
-  final trimmedString = string.trimRight();
-  final removedPortion = string.substring(trimmedString.length);
+String _tryYamlEncodeFolded(String string, int indentSize, String lineEnding) {
```

Comment:
This might be bigger than what you set out to do. But reading the code, I thinking that maybe we should restructure it a bit.


------------------

Maybe we should change how `_tryYamlEncodeFolded` and `_tryYamlEncodeLiteral` works.

What if we change the signature to:
```dart
/// Try to encode [value] as a _YAML folded string_ indent with [indent] spaces.
///
/// This method will apply an appropriate _chomping indicator_, if this is necessary
/// in order to correctly encode [value].
///
/// **Example** of a folded string.
/// ```yaml
/// key: >-
///   my folded
///   string
/// ```
///
/// See: https://yaml.org/spec/1.2.2/#813-folded-style
///
/// Returns `null`, if [value] cannot be encoded as a YAML folded string.
String? _tryYamlEncodeFolded(String value, int indent, String lineEnding) 
```

So we return `null`, when the value can't be encoded correctly as a folded string.

Then we get to change `yamlEncodeBlockScalar` such that it tries to encode in the style hinted by `YamlScalar.style`, and if this fails, we can fallback to a different block-mode, and if all block-modes fail we can fallback to quoted string mode.

This way, we don't have to write `yamlEncodeBlockScalar` such that it figures out of the value can be encoded in folded mode, and then `_tryYamlEncodeFolded` does the encoding.
Which is also weird, because `_tryYamlEncodeFolded` starts with `try`, but it actually doesn't have a mechanism for communicating back to say "oh, I can't correctly encode this value".

This way: `_tryYamlEncodeFolded` is responsible for:
 * (A) Determining if `value` can be encoded as a folded string.
 * (B) Actually encoding `value` as a folded string.

And `yamlEncodeBlockScalar` is responsible for:
 * (1) Trying `_tryYamlEncodeFolded` or `_tryYamlEncodeLiteral` (depending on `YamlScalar.style`).
 * (2) Falling back to another style if the preferred style couldn't be used.
 * (3) Falling back to using quoted string encoding (which always works), if none of the block styles could be used.

And `yamlEncodeBlockScalar` doesn't have to know what values can be encoded in what styles, it just determines the preferred option and tries it.

Reply:
I didn't see this. Lemme get to it.

Reply:
No rush :)

We can also do it in a follow up.

Btw, there values we can't encode right?
Like special characters and some weird combination of whitespace and line breaks I would also assume.

Reply:
If so, this PR is good to go. 

> No rush :)
> 
> We can also do it in a follow up.
> 
> Btw, there values we can't encode right? Like special characters and some weird combination of white space and line breaks I would also assume.

Your suggestions in the previous comment are in the ball park of what I had in mind. 

I realised the code below makes the tests pass and we never get to encode the folded string as `>+ my string` as it trims the space as the end including `\n`.

https://github.com/kekavc24/yaml_edit/blob/08a146ef8f2c7aba908d2017c9ec840b882c9e51/lib/src/strings.dart#L186-L196

> No rush :)
> 
> We can also do it in a follow up.
> 
> Btw, there values we can't encode right? Like special characters and some weird combination of white space and line breaks I would also assume.

Yeah. The trailing line breaks and white-space stand out. Your suggestions actually make it possible to correctly debug what the issue is. I'll look into it.

However, I can still tackle it in this PR. No issue. Your call.

Reply:
> Your suggestions actually make it possible to correctly debug what the issue is. I'll look into it.
>
> I can still tackle it in this PR. No issue. Your call.

It's probably a good idea to spend some time making sure we get this right :D

It'd be really nice if we didn't have so many weird string encoding bugs.